File: scripts/check-translations.py
#!/usr/bin/env python3
"""
Translation completeness checker for VibeSQL.

This script compares translations across locales to find:
- Missing message IDs (in en-US but not in target locale)
- Extra message IDs (in target locale but not in en-US)
- Variable mismatches between translations

Usage:
    python scripts/check-translations.py [--locale LOCALE] [--threshold PERCENT]

Examples:
    python scripts/check-translations.py              # Check all locales
    python scripts/check-translations.py --locale es  # Check only Spanish
    python scripts/check-translations.py --threshold 90  # Fail if < 90% coverage
"""

import argparse
import os
import re
import sys
from pathlib import Path
from typing import Dict, List, Set, Tuple, Optional
from dataclasses import dataclass


@dataclass
class MessageInfo:
    """Information about a Fluent message."""
    id: str
    variables: Set[str]
    has_attributes: bool
    line_number: int


@dataclass
class LocaleReport:
    """Report for a single locale's translation coverage."""
    locale: str
    total_reference: int
    translated: int
    missing: List[str]
    extra: List[str]
    variable_mismatches: List[Tuple[str, Set[str], Set[str]]]

    @property
    def coverage_percent(self) -> float:
        if self.total_reference == 0:
            return 100.0
        return (self.translated / self.total_reference) * 100


def find_resources_dir() -> Path:
    """Find the l10n resources directory."""
    # Try common locations
    candidates = [
        Path("crates/vibesql-l10n/resources"),
        Path("../crates/vibesql-l10n/resources"),
        Path("../../crates/vibesql-l10n/resources"),
    ]

    # Also check from script location
    script_dir = Path(__file__).parent.absolute()
    candidates.append(script_dir.parent / "crates" / "vibesql-l10n" / "resources")

    for candidate in candidates:
        if candidate.is_dir():
            return candidate.absolute()

    raise FileNotFoundError(
        "Could not find l10n resources directory. "
        "Run from repository root or specify path."
    )


def parse_ftl_file(file_path: Path) -> Dict[str, MessageInfo]:
    """
    Parse a Fluent file and extract message IDs with their variables.

    Returns a dict mapping message ID to MessageInfo.
    """
    messages = {}

    if not file_path.exists():
        return messages

    content = file_path.read_text(encoding="utf-8")
    lines = content.split("\n")

    current_msg_id = None
    current_line = 0

    # Regex patterns
    msg_pattern = re.compile(r"^([a-zA-Z][a-zA-Z0-9_-]*)\s*=")
    attr_pattern = re.compile(r"^\s+\.([a-zA-Z][a-zA-Z0-9_-]*)\s*=")
    var_pattern = re.compile(r"\{\s*\$([a-zA-Z_][a-zA-Z0-9_]*)")

    for i, line in enumerate(lines, 1):
        # Skip comments and empty lines
        if line.strip().startswith("#") or not line.strip():
            continue

        # Check for new message
        msg_match = msg_pattern.match(line)
        if msg_match:
            current_msg_id = msg_match.group(1)
            current_line = i
            variables = set(var_pattern.findall(line))
            messages[current_msg_id] = MessageInfo(
                id=current_msg_id,
                variables=variables,
                has_attributes=False,
                line_number=i
            )
            continue

        # Check for attribute (continuation of previous message)
        attr_match = attr_pattern.match(line)
        if attr_match and current_msg_id:
            messages[current_msg_id].has_attributes = True
            # Add variables from attribute line
            messages[current_msg_id].variables.update(var_pattern.findall(line))
            continue

        # Check for continuation line (starts with whitespace)
        if line.startswith(" ") or line.startswith("\t"):
            if current_msg_id and current_msg_id in messages:
                # Add variables from continuation line
                messages[current_msg_id].variables.update(var_pattern.findall(line))

    return messages


def get_all_locales(resources_dir: Path) -> List[str]:
    """Get all available locale directories."""
    locales = []
    for item in resources_dir.iterdir():
        if item.is_dir() and not item.name.startswith("."):
            locales.append(item.name)
    return sorted(locales)


def get_ftl_files(locale_dir: Path) -> List[str]:
    """Get all .ftl files in a locale directory."""
    return sorted([f.name for f in locale_dir.glob("*.ftl")])


def check_locale(
    resources_dir: Path,
    locale: str,
    reference_locale: str = "en-US"
) -> LocaleReport:
    """
    Check a locale against the reference locale.

    Returns a LocaleReport with coverage information.
    """
    ref_dir = resources_dir / reference_locale
    target_dir = resources_dir / locale

    # Collect all messages from reference locale
    ref_messages: Dict[str, MessageInfo] = {}
    for ftl_file in get_ftl_files(ref_dir):
        file_messages = parse_ftl_file(ref_dir / ftl_file)
        ref_messages.update(file_messages)

    # Collect all messages from target locale
    target_messages: Dict[str, MessageInfo] = {}
    for ftl_file in get_ftl_files(ref_dir):  # Use ref_dir to check same files
        target_path = target_dir / ftl_file
        if target_path.exists():
            file_messages = parse_ftl_file(target_path)
            target_messages.update(file_messages)

    # Find missing and extra messages
    ref_ids = set(ref_messages.keys())
    target_ids = set(target_messages.keys())

    missing = sorted(ref_ids - target_ids)
    extra = sorted(target_ids - ref_ids)

    # Check variable consistency for translated messages
    variable_mismatches = []
    for msg_id in ref_ids & target_ids:
        ref_vars = ref_messages[msg_id].variables
        target_vars = target_messages[msg_id].variables
        if ref_vars != target_vars:
            variable_mismatches.append((msg_id, ref_vars, target_vars))

    return LocaleReport(
        locale=locale,
        total_reference=len(ref_ids),
        translated=len(ref_ids) - len(missing),
        missing=missing,
        extra=extra,
        variable_mismatches=variable_mismatches
    )


def print_report(report: LocaleReport, verbose: bool = False) -> None:
    """Print a locale report."""
    coverage = report.coverage_percent

    # Color codes for terminal
    green = "\033[32m"
    yellow = "\033[33m"
    red = "\033[31m"
    reset = "\033[0m"

    # Choose color based on coverage
    if coverage >= 100:
        color = green
        status = "COMPLETE"
    elif coverage >= 80:
        color = yellow
        status = "GOOD"
    else:
        color = red
        status = "NEEDS WORK"

    print(f"\n{color}=== {report.locale} ==={reset}")
    print(f"Coverage: {color}{coverage:.1f}%{reset} ({report.translated}/{report.total_reference}) [{status}]")

    if report.missing:
        print(f"\n{yellow}Missing messages ({len(report.missing)}):{reset}")
        if verbose or len(report.missing) <= 10:
            for msg_id in report.missing:
                print(f"  - {msg_id}")
        else:
            for msg_id in report.missing[:10]:
                print(f"  - {msg_id}")
            print(f"  ... and {len(report.missing) - 10} more")

    if report.extra:
        print(f"\n{yellow}Extra messages ({len(report.extra)}):{reset}")
        if verbose or len(report.extra) <= 5:
            for msg_id in report.extra:
                print(f"  - {msg_id}")
        else:
            for msg_id in report.extra[:5]:
                print(f"  - {msg_id}")
            print(f"  ... and {len(report.extra) - 5} more")

    if report.variable_mismatches:
        print(f"\n{red}Variable mismatches ({len(report.variable_mismatches)}):{reset}")
        for msg_id, ref_vars, target_vars in report.variable_mismatches:
            missing_vars = ref_vars - target_vars
            extra_vars = target_vars - ref_vars
            print(f"  - {msg_id}:")
            if missing_vars:
                print(f"      Missing: {', '.join(sorted(missing_vars))}")
            if extra_vars:
                print(f"      Extra: {', '.join(sorted(extra_vars))}")


def generate_badge_json(reports: List[LocaleReport], output_path: Optional[Path] = None) -> dict:
    """Generate a JSON object for shields.io badge."""
    # Calculate average coverage
    if not reports:
        avg_coverage = 0
    else:
        avg_coverage = sum(r.coverage_percent for r in reports) / len(reports)

    # Determine color
    if avg_coverage >= 100:
        color = "brightgreen"
    elif avg_coverage >= 90:
        color = "green"
    elif avg_coverage >= 80:
        color = "yellow"
    elif avg_coverage >= 50:
        color = "orange"
    else:
        color = "red"

    badge = {
        "schemaVersion": 1,
        "label": "i18n",
        "message": f"{avg_coverage:.0f}%",
        "color": color
    }

    if output_path:
        import json
        output_path.write_text(json.dumps(badge, indent=2))

    return badge


def main():
    parser = argparse.ArgumentParser(
        description="Check translation completeness for VibeSQL"
    )
    parser.add_argument(
        "--locale",
        help="Check only this locale (default: all)"
    )
    parser.add_argument(
        "--threshold",
        type=float,
        default=0,
        help="Minimum coverage percentage required (default: 0, no threshold)"
    )
    parser.add_argument(
        "--verbose", "-v",
        action="store_true",
        help="Show all missing messages, not just first 10"
    )
    parser.add_argument(
        "--badge",
        type=Path,
        help="Output path for shields.io badge JSON"
    )
    parser.add_argument(
        "--ci",
        action="store_true",
        help="CI mode: exit with error if any locale is below threshold"
    )

    args = parser.parse_args()

    try:
        resources_dir = find_resources_dir()
    except FileNotFoundError as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

    print(f"Resources directory: {resources_dir}")

    # Get locales to check
    all_locales = get_all_locales(resources_dir)
    reference_locale = "en-US"

    if reference_locale not in all_locales:
        print(f"Error: Reference locale '{reference_locale}' not found", file=sys.stderr)
        sys.exit(1)

    # Filter locales if specific one requested
    if args.locale:
        if args.locale not in all_locales:
            print(f"Error: Locale '{args.locale}' not found", file=sys.stderr)
            print(f"Available locales: {', '.join(all_locales)}")
            sys.exit(1)
        locales_to_check = [args.locale]
    else:
        locales_to_check = [l for l in all_locales if l != reference_locale]

    if not locales_to_check:
        print("No locales to check (only reference locale exists)")
        sys.exit(0)

    print(f"Checking {len(locales_to_check)} locale(s) against {reference_locale}")

    # Check each locale
    reports = []
    below_threshold = []

    for locale in locales_to_check:
        report = check_locale(resources_dir, locale, reference_locale)
        reports.append(report)
        print_report(report, args.verbose)

        if args.threshold > 0 and report.coverage_percent < args.threshold:
            below_threshold.append(locale)

    # Generate badge if requested
    if args.badge:
        badge = generate_badge_json(reports, args.badge)
        print(f"\nBadge JSON written to: {args.badge}")

    # Summary
    print("\n" + "=" * 50)
    print("SUMMARY")
    print("=" * 50)

    for report in reports:
        status = "OK" if report.coverage_percent >= args.threshold else "BELOW THRESHOLD"
        print(f"{report.locale}: {report.coverage_percent:.1f}% [{status}]")

    # Exit with error in CI mode if below threshold
    if args.ci and below_threshold:
        print(f"\nError: {len(below_threshold)} locale(s) below {args.threshold}% threshold:")
        for locale in below_threshold:
            print(f"  - {locale}")
        sys.exit(1)

    # Exit with error if there are variable mismatches
    total_mismatches = sum(len(r.variable_mismatches) for r in reports)
    if total_mismatches > 0:
        print(f"\nWarning: {total_mismatches} variable mismatch(es) found")
        if args.ci:
            sys.exit(1)

    print("\nDone!")


if __name__ == "__main__":
    main()

File: crates/vibesql-l10n/src/bin/check-translations.rs
//! Translation Completeness Checker
//!
//! This binary validates that all locales have complete translations compared to en-US.
//! It checks for:
//! - Missing message IDs in non-English locales
//! - Missing .ftl files in non-English locales
//! - Extra/obsolete messages not in en-US (warnings)
//!
//! Usage:
//!     cargo run -p vibesql-l10n --bin check-translations
//!     cargo run -p vibesql-l10n --bin check-translations -- --verbose
//!     cargo run -p vibesql-l10n --bin check-translations -- --locale es
//!     cargo run -p vibesql-l10n --bin check-translations -- --json
//!
//! Exit codes:
//!     0 - All translations complete
//!     1 - Missing translations found

use fluent_syntax::parser::parse;
use std::collections::{BTreeMap, BTreeSet};
use std::path::{Path, PathBuf};
use std::process::ExitCode;

/// ANSI color codes for terminal output
mod colors {
    pub const RED: &str = "\x1b[31m";
    pub const GREEN: &str = "\x1b[32m";
    pub const YELLOW: &str = "\x1b[33m";
    pub const CYAN: &str = "\x1b[36m";
    pub const BOLD: &str = "\x1b[1m";
    pub const RESET: &str = "\x1b[0m";
}

/// Command-line arguments
struct Args {
    verbose: bool,
    json: bool,
    locale: Option<String>,
    strict: bool,
}

impl Args {
    fn parse() -> Self {
        let args: Vec<String> = std::env::args().collect();
        let mut verbose = false;
        let mut json = false;
        let mut locale = None;
        let mut strict = false;

        let mut i = 1;
        while i < args.len() {
            match args[i].as_str() {
                "--verbose" | "-v" => verbose = true,
                "--json" => json = true,
                "--strict" => strict = true,
                "--locale" | "-l" => {
                    i += 1;
                    if i < args.len() {
                        locale = Some(args[i].clone());
                    }
                }
                "--help" | "-h" => {
                    print_help();
                    std::process::exit(0);
                }
                _ => {}
            }
            i += 1;
        }

        Self {
            verbose,
            json,
            locale,
            strict,
        }
    }
}

fn print_help() {
    println!(
        r#"Translation Completeness Checker

USAGE:
    cargo run -p vibesql-l10n --bin check-translations [OPTIONS]

OPTIONS:
    -v, --verbose    Show all messages, including present ones
    -l, --locale     Check only a specific locale (e.g., --locale es)
        --json       Output results as JSON
        --strict     Treat warnings (extra messages) as errors
    -h, --help       Print this help message

EXAMPLES:
    # Check all locales
    cargo run -p vibesql-l10n --bin check-translations

    # Check Spanish translations only
    cargo run -p vibesql-l10n --bin check-translations -- --locale es

    # Verbose output showing all messages
    cargo run -p vibesql-l10n --bin check-translations -- --verbose

    # Output as JSON for CI integration
    cargo run -p vibesql-l10n --bin check-translations -- --json
"#
    );
}

/// Result for a single locale
#[derive(Default)]
struct LocaleResult {
    locale: String,
    missing_files: Vec<String>,
    missing_messages: BTreeMap<String, Vec<String>>, // file -> [message_ids]
    extra_messages: BTreeMap<String, Vec<String>>,   // file -> [message_ids]
    total_expected: usize,
    total_present: usize,
}

impl LocaleResult {
    fn coverage_percent(&self) -> f64 {
        if self.total_expected == 0 {
            100.0
        } else {
            (self.total_present as f64 / self.total_expected as f64) * 100.0
        }
    }

    fn has_issues(&self) -> bool {
        !self.missing_files.is_empty()
            || self.missing_messages.values().any(|v| !v.is_empty())
    }

    fn has_warnings(&self) -> bool {
        self.extra_messages.values().any(|v| !v.is_empty())
    }
}

/// Find the resources directory relative to the crate
fn find_resources_dir() -> Option<PathBuf> {
    let candidates = [
        PathBuf::from("crates/vibesql-l10n/resources"),
        PathBuf::from("resources"),
        PathBuf::from("../../../crates/vibesql-l10n/resources"),
    ];

    candidates.into_iter().find(|c| c.is_dir())
}

/// Get all locale directories in the resources directory
fn get_locales(resources_dir: &Path) -> Vec<String> {
    let mut locales = Vec::new();

    if let Ok(entries) = std::fs::read_dir(resources_dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_dir() {
                if let Some(name) = path.file_name() {
                    locales.push(name.to_string_lossy().to_string());
                }
            }
        }
    }

    locales.sort();
    locales
}

/// Get all .ftl files in a directory (returns just filenames)
fn get_ftl_files(dir: &Path) -> Vec<String> {
    let mut files = Vec::new();

    if let Ok(entries) = std::fs::read_dir(dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.extension().is_some_and(|e| e == "ftl") {
                if let Some(name) = path.file_name() {
                    files.push(name.to_string_lossy().to_string());
                }
            }
        }
    }

    files.sort();
    files
}

/// Extract all message IDs from a .ftl file
fn extract_message_ids(path: &Path) -> BTreeSet<String> {
    let mut ids = BTreeSet::new();

    let content = match std::fs::read_to_string(path) {
        Ok(c) => c,
        Err(_) => return ids,
    };

    let resource = match parse(content.as_str()) {
        Ok(r) => r,
        Err((r, _errors)) => r,
    };

    for entry in &resource.body {
        match entry {
            fluent_syntax::ast::Entry::Message(msg) => {
                ids.insert(msg.id.name.to_string());
            }
            fluent_syntax::ast::Entry::Term(term) => {
                ids.insert(format!("-{}", term.id.name));
            }
            _ => {}
        }
    }

    ids
}

/// Check a single locale against en-US
fn check_locale(
    resources_dir: &Path,
    locale: &str,
    en_us_messages: &BTreeMap<String, BTreeSet<String>>,
) -> LocaleResult {
    let mut result = LocaleResult {
        locale: locale.to_string(),
        ..Default::default()
    };

    let locale_dir = resources_dir.join(locale);

    // Check for missing files
    for file in en_us_messages.keys() {
        let file_path = locale_dir.join(file);
        if !file_path.exists() {
            result.missing_files.push(file.clone());
            // All messages in this file are missing
            if let Some(msgs) = en_us_messages.get(file) {
                result
                    .missing_messages
                    .insert(file.clone(), msgs.iter().cloned().collect());
                result.total_expected += msgs.len();
            }
        } else {
            // File exists, check for missing messages
            let locale_ids = extract_message_ids(&file_path);
            if let Some(en_ids) = en_us_messages.get(file) {
                result.total_expected += en_ids.len();

                let missing: Vec<String> = en_ids.difference(&locale_ids).cloned().collect();
                let extra: Vec<String> = locale_ids.difference(en_ids).cloned().collect();

                result.total_present += en_ids.len() - missing.len();

                if !missing.is_empty() {
                    result.missing_messages.insert(file.clone(), missing);
                }
                if !extra.is_empty() {
                    result.extra_messages.insert(file.clone(), extra);
                }
            }
        }
    }

    result
}

/// Print results in human-readable format
fn print_results(results: &[LocaleResult], verbose: bool) {
    println!(
        "\n{}Translation Completeness Report{}",
        colors::BOLD,
        colors::RESET
    );
    println!("================================\n");

    for result in results {
        let coverage = result.coverage_percent();
        let color = if coverage >= 100.0 {
            colors::GREEN
        } else if coverage >= 80.0 {
            colors::YELLOW
        } else {
            colors::RED
        };

        println!(
            "{}{}{}  {:.1}% ({}/{} messages)",
            color,
            result.locale,
            colors::RESET,
            coverage,
            result.total_present,
            result.total_expected
        );

        // Missing files
        if !result.missing_files.is_empty() {
            println!(
                "  {}Missing files:{}",
                colors::RED,
                colors::RESET
            );
            for file in &result.missing_files {
                println!("    - {}", file);
            }
        }

        // Missing messages
        for (file, messages) in &result.missing_messages {
            if !messages.is_empty() && !result.missing_files.contains(file) {
                println!(
                    "  {}Missing in {}:{} ({} messages)",
                    colors::RED,
                    file,
                    colors::RESET,
                    messages.len()
                );
                if verbose {
                    for msg in messages {
                        println!("    - {}", msg);
                    }
                }
            }
        }

        // Extra messages (warnings)
        if result.has_warnings() {
            for (file, messages) in &result.extra_messages {
                if !messages.is_empty() {
                    println!(
                        "  {}Extra in {}:{} ({} messages, not in en-US)",
                        colors::YELLOW,
                        file,
                        colors::RESET,
                        messages.len()
                    );
                    if verbose {
                        for msg in messages {
                            println!("    - {}", msg);
                        }
                    }
                }
            }
        }

        println!();
    }
}

/// Print results in JSON format
fn print_json(results: &[LocaleResult]) {
    println!("{{");
    println!("  \"locales\": [");
    for (i, result) in results.iter().enumerate() {
        println!("    {{");
        println!("      \"locale\": \"{}\",", result.locale);
        println!("      \"coverage\": {:.2},", result.coverage_percent());
        println!("      \"total_expected\": {},", result.total_expected);
        println!("      \"total_present\": {},", result.total_present);
        println!("      \"missing_files\": {:?},", result.missing_files);

        // Missing messages
        println!("      \"missing_messages\": {{");
        let missing_entries: Vec<_> = result
            .missing_messages
            .iter()
            .filter(|(_, v)| !v.is_empty())
            .collect();
        for (j, (file, messages)) in missing_entries.iter().enumerate() {
            let comma = if j < missing_entries.len() - 1 { "," } else { "" };
            println!("        \"{}\": {:?}{}", file, messages, comma);
        }
        println!("      }},");

        // Extra messages
        println!("      \"extra_messages\": {{");
        let extra_entries: Vec<_> = result
            .extra_messages
            .iter()
            .filter(|(_, v)| !v.is_empty())
            .collect();
        for (j, (file, messages)) in extra_entries.iter().enumerate() {
            let comma = if j < extra_entries.len() - 1 { "," } else { "" };
            println!("        \"{}\": {:?}{}", file, messages, comma);
        }
        println!("      }}");

        let comma = if i < results.len() - 1 { "," } else { "" };
        println!("    }}{}", comma);
    }
    println!("  ]");
    println!("}}");
}

fn main() -> ExitCode {
    let args = Args::parse();

    // Find resources directory
    let resources_dir = match find_resources_dir() {
        Some(dir) => dir,
        None => {
            eprintln!(
                "{}Error: Could not find resources directory{}",
                colors::RED,
                colors::RESET
            );
            eprintln!("Run from workspace root or crate directory.");
            return ExitCode::FAILURE;
        }
    };

    if !args.json {
        println!(
            "{}Checking translation completeness...{}",
            colors::CYAN,
            colors::RESET
        );
        println!("Resources directory: {}\n", resources_dir.display());
    }

    // Get en-US as the source of truth
    let en_us_dir = resources_dir.join("en-US");
    if !en_us_dir.exists() {
        eprintln!(
            "{}Error: en-US locale directory not found{}",
            colors::RED,
            colors::RESET
        );
        return ExitCode::FAILURE;
    }

    // Build en-US message catalog
    let mut en_us_messages: BTreeMap<String, BTreeSet<String>> = BTreeMap::new();
    for file in get_ftl_files(&en_us_dir) {
        let ids = extract_message_ids(&en_us_dir.join(&file));
        en_us_messages.insert(file, ids);
    }

    let total_en_messages: usize = en_us_messages.values().map(|s| s.len()).sum();
    if !args.json {
        println!(
            "Reference locale (en-US): {} files, {} messages\n",
            en_us_messages.len(),
            total_en_messages
        );
    }

    // Get all locales to check
    let locales: Vec<String> = if let Some(ref locale) = args.locale {
        vec![locale.clone()]
    } else {
        get_locales(&resources_dir)
            .into_iter()
            .filter(|l| l != "en-US")
            .collect()
    };

    // Check each locale
    let mut results: Vec<LocaleResult> = Vec::new();
    for locale in &locales {
        let result = check_locale(&resources_dir, locale, &en_us_messages);
        results.push(result);
    }

    // Output results
    if args.json {
        print_json(&results);
    } else {
        print_results(&results, args.verbose);

        // Summary
        println!("{}Summary{}", colors::BOLD, colors::RESET);
        println!("=======");

        let complete_count = results.iter().filter(|r| !r.has_issues()).count();
        let total_missing: usize = results
            .iter()
            .flat_map(|r| r.missing_messages.values())
            .map(|v| v.len())
            .sum();

        println!("Locales checked: {}", results.len());
        println!(
            "Complete locales: {}{}/{}{}",
            if complete_count == results.len() {
                colors::GREEN
            } else {
                colors::YELLOW
            },
            complete_count,
            results.len(),
            colors::RESET
        );
        println!(
            "Total missing messages: {}{}{}",
            if total_missing > 0 {
                colors::RED
            } else {
                colors::GREEN
            },
            total_missing,
            colors::RESET
        );
    }

    // Determine exit code
    let has_missing = results.iter().any(|r| r.has_issues());
    let has_extra = results.iter().any(|r| r.has_warnings());

    if has_missing || (args.strict && has_extra) {
        if !args.json {
            println!(
                "\n{}INCOMPLETE: Missing translations found{}",
                colors::RED,
                colors::RESET
            );
        }
        ExitCode::FAILURE
    } else {
        if !args.json {
            println!(
                "\n{}COMPLETE: All translations present{}",
                colors::GREEN,
                colors::RESET
            );
        }
        ExitCode::SUCCESS
    }
}
